pkg/cloud/talos/provisioners/azure/azure.go

```diff
@@ -0,0 +1,416 @@
+package azure
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-04-01/network"
+	"github.com/Azure/go-autorest/autorest"
+	azuresdk "github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/talos-systems/cluster-api-provider-talos/pkg/cloud/talos/utils"
+	"gopkg.in/yaml.v2"
+	"k8s.io/client-go/kubernetes"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+// Az represents a provider for Azure.
+type Az struct {
+}
+
+// ClusterInfo holds data about desired config in cluster object
+type ClusterInfo struct {
+	Location      string
+	ResourceGroup string
+	Instances     InstanceInfo
+}
+
+// InstanceInfo holds data about the instances we'll create
+type InstanceInfo struct {
+	Type    string
+	Image   string
+	Network string
+	Subnet  string
+	Disks   DiskInfo
+}
+
+// DiskInfo holds disk info data
+type DiskInfo struct {
+	Size int
+}
+
+// Session is an object representing session for subscription
+type Session struct {
+	SubscriptionID string
+	Authorizer     autorest.Authorizer
+}
+
+// NewAz returns an instance of the Azure provisioner
+func NewAz() (*Az, error) {
+	return &Az{}, nil
+}
+
+// Create creates an instance in Azure.
+func (azure *Az) Create(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, clientset *kubernetes.Clientset) error {
+	clusterSpec, err := utils.ClusterProviderFromSpec(cluster.Spec.ProviderSpec)
+	if err != nil {
+		return err
+	}
+
+	machineSpec, err := utils.MachineProviderFromSpec(machine.Spec.ProviderSpec)
+	if err != nil {
+		return err
+	}
+
+	azureConfig := &ClusterInfo{}
+	yaml.Unmarshal([]byte(machineSpec.Platform.Config), azureConfig)
+
+	// Dig out the subnet for our nic
+	subnet, err := getSubnetByName(ctx, azureConfig)
+	if err != nil {
+		return err
+	}
+
+	// Begin crafting the config for the nic
+	nicIPConfigProperties := &network.InterfaceIPConfigurationPropertiesFormat{
+		PrivateIPAllocationMethod: network.Dynamic,
+		Subnet:                    subnet,
+	}
+
+	// Find the public IP we want to use if necessary
+	if !strings.Contains(machine.ObjectMeta.Name, "worker") {
+		nameSlice := strings.Split(machine.ObjectMeta.Name, "-")
+		indexString := nameSlice[len(nameSlice)-1]
+		index, err := strconv.Atoi(indexString)
+		if err != nil {
+			return err
+		}
+		publicIP := clusterSpec.Masters.IPs[index]
+		publicIPObject, err := getPublicIPbyIP(ctx, publicIP)
```

Comment:
`getPublicIPbyIP` can return `nil, nil` which is not checked? This would result in NIC with no Public IP if I'm not mistaken.